src/cmd/promremotebench/query.go

```diff
@@ -63,25 +70,105 @@ func newQueryExecutor(opts queryExecutorOptions) *queryExecutor {
 	}
 }
 
-func (q *queryExecutor) Run() {
+func (q *queryExecutor) Run(checker Checker) {
 	q.Logger.Info("query load configured",
 		zap.Int("concurrency", q.Concurrency))
 	for i := 0; i < q.Concurrency; i++ {
-		go q.workerLoop()
+		go q.alertLoad(checker)
 	}
+
+	go q.accuracyCheck(checker)
 }
 
-func (q *queryExecutor) workerLoop() {
+// accuracyCheck checks the accuracy of data for one
+// host at a time.
+func (q *queryExecutor) accuracyCheck(checker Checker) {
+	type label struct {
+		name  string
+		value string
+	}
+	labels := make([]label, 0, len(q.Labels))
+	for k, v := range q.Labels {
+		labels = append(labels, label{name: k, value: v})
+	}
+
+	query := new(strings.Builder)
+	for i := 0; ; i++ {
+		func() {
+			// Exec in func to be able to use defer.
+			if i > 0 {
+				time.Sleep(q.Sleep)
+			}
+
+			query.Reset()
+			if q.Aggregation != "" {
+				mustWriteString(query, q.Aggregation)
+				mustWriteString(query, "({")
+			}
+
+			curHostnames := checker.GetHostNames()
+			if len(curHostnames) == 0 {
+				if i > 0 {
+					q.Logger.Error("no hosts returned in the checker, skipping accuracy check.")
+				}
+				return
+			}
+
+			var (
+				dps          []Datapoint
+				selectedHost string
+			)
+
+			for j := 0; j < 5; j++ {
+				selectedHost = curHostnames[rand.Intn(len(curHostnames))]
+				dps = checker.GetDatapoints(selectedHost)
+				if len(dps) > 1 {
+					break
+				}
+			}
+
+			if len(dps) <= 1 && i > 1 {
+				q.Logger.Error("couldn't find a host with more than 1 datapoint. Skipping accuracy check")
+			}
+
+			mustWriteString(query, "hostname=\""+selectedHost+"\"")
+
+			// Write the common labels.
+			for j := 0; j < len(labels); j++ {
+				mustWriteString(query, ",")
+
+				l := labels[j]
+				mustWriteString(query, l.name)
+				mustWriteString(query, "=\"")
+				mustWriteString(query, l.value)
+				mustWriteString(query, "\"")
+			}
+
+			if q.Aggregation != "" {
+				mustWriteString(query, "})")
+			}
+
+			res := q.executeQuery(query, true, q.AccuracyRange, q.AccuracyStep)
+			if len(res) == 0 {
+				q.Logger.Error("invalid response for accuracy query")
+			}
+
+			q.validateQuery(dps, res)
```

Comment:
Do you want to do anything with the bool that gets returned?

Reply:
Not really, all errors are logged in the func itself, I made it return mostly for tests.